File: server/mfup/protocol.py
"""MFUP/2 protocol types and binary frame decoder (server-side)."""

from __future__ import annotations

import enum
import struct
from dataclasses import dataclass
from typing import Optional


PROTOCOL_VERSION = "MFUP/2"
ROOT_NODE_ID = 0


# ---------------------------------------------------------------------------
# Frame tags
# ---------------------------------------------------------------------------
class FrameTag(enum.IntEnum):
    NODE = 0x01
    SUMMARY = 0x02
    FILE_OPEN = 0x03
    FILE_CHUNK = 0x04
    FILE_CLOSE = 0x05
    DIR_CLOSE = 0x06
    SESSION_END = 0x07
    CLIENT_ABORT = 0x08


class NodeKind(enum.IntEnum):
    DIR = 0x00
    FILE = 0x01


class ChecksumKind(enum.IntEnum):
    CRC32C = 0x01


class HashKind(enum.IntEnum):
    NONE = 0x00
    SHA256 = 0x01


# ---------------------------------------------------------------------------
# Decoded frame dataclasses
# ---------------------------------------------------------------------------
@dataclass(slots=True)
class NodeFrame:
    tag: FrameTag = FrameTag.NODE
    node_id: int = 0
    parent_id: int = 0
    kind: NodeKind = NodeKind.FILE
    name: str = ""
    size_hint: Optional[int] = None
    mtime_ms: Optional[int] = None


@dataclass(slots=True)
class SummaryFrame:
    tag: FrameTag = FrameTag.SUMMARY
    node_id: int = 0
    scan_done_units: int = 0
    scan_est_units: int = 0
    body_done_bytes: int = 0
    body_est_bytes: int = 0
    sealed: bool = False


@dataclass(slots=True)
class FileOpenFrame:
    tag: FrameTag = FrameTag.FILE_OPEN
    node_id: int = 0
    size: int = 0
    mtime_ms: Optional[int] = None


@dataclass(slots=True)
class FileChunkFrame:
    tag: FrameTag = FrameTag.FILE_CHUNK
    node_id: int = 0
    offset: int = 0
    length: int = 0
    checksum_kind: ChecksumKind = ChecksumKind.CRC32C
    checksum: int = 0
    payload: bytes = b""


@dataclass(slots=True)
class FileCloseFrame:
    tag: FrameTag = FrameTag.FILE_CLOSE
    node_id: int = 0
    size_sent: int = 0
    strong_hash_kind: HashKind = HashKind.NONE
    strong_hash: Optional[bytes] = None


@dataclass(slots=True)
class DirCloseFrame:
    tag: FrameTag = FrameTag.DIR_CLOSE
    node_id: int = 0


@dataclass(slots=True)
class SessionEndFrame:
    tag: FrameTag = FrameTag.SESSION_END
    scan_done_units: int = 0
    scan_est_units: int = 0
    body_done_bytes: int = 0
    body_est_bytes: int = 0
    sealed: bool = True


@dataclass(slots=True)
class ClientAbortFrame:
    tag: FrameTag = FrameTag.CLIENT_ABORT
    code: str = ""
    reason: str = ""


Frame = (
    NodeFrame
    | SummaryFrame
    | FileOpenFrame
    | FileChunkFrame
    | FileCloseFrame
    | DirCloseFrame
    | SessionEndFrame
    | ClientAbortFrame
)


# Server → Client control message: {"t": "PROBE_ACK"}


# ---------------------------------------------------------------------------
# Session states
# ---------------------------------------------------------------------------
class SessionState(str, enum.Enum):
    ACTIVE = "active"
    PAUSED_BY_SERVER = "paused_by_server"
    WAITING_RESUME = "waiting_resume"
    COMMITTING = "committing"
    COMMITTED = "committed"
    ABORTED = "aborted"
    EXPIRED = "expired"
    FAILED = "failed"


# ---------------------------------------------------------------------------
# Node statuses (in DB)
# ---------------------------------------------------------------------------
class NodeStatus(str, enum.Enum):
    OPEN = "open"
    CLOSED = "closed"
    REJECTED = "rejected"
    PRUNED = "pruned"


# ---------------------------------------------------------------------------
# Binary frame decoder
#
# Wire format: [4-byte big-endian length][1-byte tag][payload]
# The length covers tag + payload.
# ---------------------------------------------------------------------------

def _read_u8(data: memoryview, off: int) -> tuple[int, int]:
    return data[off], off + 1


def _read_u16(data: memoryview, off: int) -> tuple[int, int]:
    return struct.unpack_from("!H", data, off)[0], off + 2


def _read_u32(data: memoryview, off: int) -> tuple[int, int]:
    return struct.unpack_from("!I", data, off)[0], off + 4


def _read_u64(data: memoryview, off: int) -> tuple[int, int]:
    return struct.unpack_from("!Q", data, off)[0], off + 8


def _read_string(data: memoryview, off: int) -> tuple[str, int]:
    slen, off = _read_u16(data, off)
    s = bytes(data[off : off + slen]).decode("utf-8")
    return s, off + slen


def decode_frame_payload(tag: int, payload: memoryview) -> Frame:
    """Decode the payload bytes (after the tag byte) into a Frame dataclass."""
    off = 0

    if tag == FrameTag.NODE:
        node_id, off = _read_u32(payload, off)
        parent_id, off = _read_u32(payload, off)
        kind_val, off = _read_u8(payload, off)
        name, off = _read_string(payload, off)
        has_size, off = _read_u8(payload, off)
        size_hint = None
        if has_size:
            size_hint, off = _read_u64(payload, off)
        has_mtime, off = _read_u8(payload, off)
        mtime_ms = None
        if has_mtime:
            mtime_ms, off = _read_u64(payload, off)
        return NodeFrame(
            node_id=node_id,
            parent_id=parent_id,
            kind=NodeKind(kind_val),
            name=name,
            size_hint=size_hint,
            mtime_ms=mtime_ms,
        )

    elif tag == FrameTag.SUMMARY:
        node_id, off = _read_u32(payload, off)
        scan_done, off = _read_u64(payload, off)
        scan_est, off = _read_u64(payload, off)
        body_done, off = _read_u64(payload, off)
        body_est, off = _read_u64(payload, off)
        sealed_val, off = _read_u8(payload, off)
        return SummaryFrame(
            node_id=node_id,
            scan_done_units=scan_done,
            scan_est_units=scan_est,
            body_done_bytes=body_done,
            body_est_bytes=body_est,
            sealed=bool(sealed_val),
        )

    elif tag == FrameTag.FILE_OPEN:
        node_id, off = _read_u32(payload, off)
        size, off = _read_u64(payload, off)
        has_mtime, off = _read_u8(payload, off)
        mtime_ms = None
        if has_mtime:
            mtime_ms, off = _read_u64(payload, off)
        return FileOpenFrame(node_id=node_id, size=size, mtime_ms=mtime_ms)

    elif tag == FrameTag.FILE_CHUNK:
        node_id, off = _read_u32(payload, off)
        offset_val, off = _read_u64(payload, off)
        length, off = _read_u32(payload, off)
        ck_kind, off = _read_u8(payload, off)
        checksum, off = _read_u32(payload, off)
        data = bytes(payload[off : off + length])
        return FileChunkFrame(
            node_id=node_id,
            offset=offset_val,
            length=length,
            checksum_kind=ChecksumKind(ck_kind),
            checksum=checksum,
            payload=data,
        )

    elif tag == FrameTag.FILE_CLOSE:
        node_id, off = _read_u32(payload, off)
        size_sent, off = _read_u64(payload, off)
        hash_kind_val, off = _read_u8(payload, off)
        strong_hash = None
        if hash_kind_val != HashKind.NONE:
            hash_len, off = _read_u16(payload, off)
            strong_hash = bytes(payload[off : off + hash_len])
            off += hash_len
        return FileCloseFrame(
            node_id=node_id,
            size_sent=size_sent,
            strong_hash_kind=HashKind(hash_kind_val),
            strong_hash=strong_hash,
        )

    elif tag == FrameTag.DIR_CLOSE:
        node_id, off = _read_u32(payload, off)
        return DirCloseFrame(node_id=node_id)

    elif tag == FrameTag.SESSION_END:
        scan_done, off = _read_u64(payload, off)
        scan_est, off = _read_u64(payload, off)
        body_done, off = _read_u64(payload, off)
        body_est, off = _read_u64(payload, off)
        sealed_val, off = _read_u8(payload, off)
        return SessionEndFrame(
            scan_done_units=scan_done,
            scan_est_units=scan_est,
            body_done_bytes=body_done,
            body_est_bytes=body_est,
            sealed=bool(sealed_val),
        )

    elif tag == FrameTag.CLIENT_ABORT:
        code, off = _read_string(payload, off)
        reason, off = _read_string(payload, off)
        return ClientAbortFrame(code=code, reason=reason)

    else:
        raise ValueError(f"unknown frame tag: {tag:#04x}")


class FrameReader:
    """Incremental frame reader that buffers partial data from a streaming body.

    Feed chunks via `feed(data)` and iterate decoded frames via `drain()`.
    """

    def __init__(self) -> None:
        self._buf = bytearray()

    def feed(self, data: bytes) -> None:
        self._buf.extend(data)

    def drain(self) -> list[Frame]:
        frames: list[Frame] = []
        pos = 0
        total = len(self._buf)

        while pos + 4 <= total:
            frame_len = struct.unpack_from("!I", self._buf, pos)[0]
            if pos + 4 + frame_len > total:
                break  # incomplete frame
            tag = self._buf[pos + 4]
            # Copy payload to a bytes object so no memoryview holds the buffer
            payload_bytes = bytes(self._buf[pos + 5 : pos + 4 + frame_len])
            frames.append(decode_frame_payload(tag, memoryview(payload_bytes)))
            pos += 4 + frame_len

        if pos > 0:
            del self._buf[:pos]

        return frames


# ---------------------------------------------------------------------------
# CRC-32C (Castagnoli) for checksum verification
# ---------------------------------------------------------------------------

_CRC32C_TABLE: list[int] = []


def _init_crc32c_table() -> None:
    for i in range(256):
        crc = i
        for _ in range(8):
            if crc & 1:
                crc = (crc >> 1) ^ 0x82F63B78
            else:
                crc >>= 1
        _CRC32C_TABLE.append(crc & 0xFFFFFFFF)


_init_crc32c_table()


def crc32c(data: bytes | memoryview, initial: int = 0) -> int:
    crc = (initial ^ 0xFFFFFFFF) & 0xFFFFFFFF
    for b in data:
        crc = (crc >> 8) ^ _CRC32C_TABLE[(crc ^ b) & 0xFF]
    return (crc ^ 0xFFFFFFFF) & 0xFFFFFFFF

File: packages/server/src/protocol.ts
/**
 * MFUP/2 protocol types and binary frame decoder (server-side).
 *
 * Wire format: [4-byte big-endian length][1-byte tag][payload]
 * The length covers tag + payload. All multi-byte integers are big-endian;
 * strings are [u16 length][UTF-8 bytes].
 *
 * This mirrors server/mfup-core/mfup_core/protocol.py byte for byte.
 */

export const PROTOCOL_VERSION = "MFUP/2";
export const ROOT_NODE_ID = 0;

// ---------------------------------------------------------------------------
// Frame tags — plain const objects (isolatedModules-safe), same as the client.
// ---------------------------------------------------------------------------
export const FrameTag = {
  NODE: 0x01,
  SUMMARY: 0x02,
  FILE_OPEN: 0x03,
  FILE_CHUNK: 0x04,
  FILE_CLOSE: 0x05,
  DIR_CLOSE: 0x06,
  SESSION_END: 0x07,
  CLIENT_ABORT: 0x08,
} as const;
export type FrameTag = (typeof FrameTag)[keyof typeof FrameTag];

export const NodeKind = {
  DIR: 0x00,
  FILE: 0x01,
} as const;
export type NodeKind = (typeof NodeKind)[keyof typeof NodeKind];

export const ChecksumKind = {
  CRC32C: 0x01,
} as const;
export type ChecksumKind = (typeof ChecksumKind)[keyof typeof ChecksumKind];

// ---------------------------------------------------------------------------
// Decoded frame shapes
//
// Sizes/offsets are plain numbers: JS numbers are exact to 2^53-1 (~9 PB),
// far beyond any real payload, and SQLite/fs APIs consume numbers. A u64 on
// the wire that exceeds Number.MAX_SAFE_INTEGER throws at decode.
// ---------------------------------------------------------------------------
export interface NodeFrame {
  tag: typeof FrameTag.NODE;
  nodeId: number;
  parentId: number;
  kind: NodeKind;
  name: string;
  sizeHint: number | null;
  mtimeMs: number | null;
}

export interface SummaryFrame {
  tag: typeof FrameTag.SUMMARY;
  nodeId: number;
  scanDoneUnits: number;
  scanEstUnits: number;
  bodyDoneBytes: number;
  bodyEstBytes: number;
  sealed: boolean;
}

export interface FileOpenFrame {
  tag: typeof FrameTag.FILE_OPEN;
  nodeId: number;
  size: number;
  mtimeMs: number | null;
}

export interface FileChunkFrame {
  tag: typeof FrameTag.FILE_CHUNK;
  nodeId: number;
  offset: number;
  length: number;
  checksumKind: ChecksumKind;
  checksum: number;
  payload: Uint8Array;
}

export interface FileCloseFrame {
  tag: typeof FrameTag.FILE_CLOSE;
  nodeId: number;
  sizeSent: number;
}

export interface DirCloseFrame {
  tag: typeof FrameTag.DIR_CLOSE;
  nodeId: number;
}

export interface SessionEndFrame {
  tag: typeof FrameTag.SESSION_END;
  scanDoneUnits: number;
  scanEstUnits: number;
  bodyDoneBytes: number;
  bodyEstBytes: number;
  sealed: boolean;
}

export interface ClientAbortFrame {
  tag: typeof FrameTag.CLIENT_ABORT;
  code: string;
  reason: string;
}

export type Frame =
  | NodeFrame
  | SummaryFrame
  | FileOpenFrame
  | FileChunkFrame
  | FileCloseFrame
  | DirCloseFrame
  | SessionEndFrame
  | ClientAbortFrame;

// ---------------------------------------------------------------------------
// Session states (persisted as TEXT in the per-session SQLite journal)
// ---------------------------------------------------------------------------
export const SessionState = {
  ACTIVE: "active",
  PAUSED_BY_SERVER: "paused_by_server",
  WAITING_RESUME: "waiting_resume",
  COMMITTING: "committing",
  COMMITTED: "committed",
  ABORTED: "aborted",
  EXPIRED: "expired",
  FAILED: "failed",
} as const;
export type SessionState = (typeof SessionState)[keyof typeof SessionState];

export const NodeStatus = {
  OPEN: "open",
  CLOSED: "closed",
  REJECTED: "rejected",
  PRUNED: "pruned",
} as const;
export type NodeStatus = (typeof NodeStatus)[keyof typeof NodeStatus];

// ---------------------------------------------------------------------------
// Binary frame decoder
// ---------------------------------------------------------------------------

const TEXT_DECODER = new TextDecoder("utf-8", { fatal: true });

class Cursor {
  off = 0;
  constructor(
    readonly view: DataView,
    readonly bytes: Uint8Array,
  ) {}

  u8(): number {
    const v = this.view.getUint8(this.off);
    this.off += 1;
    return v;
  }
  u16(): number {
    const v = this.view.getUint16(this.off);
    this.off += 2;
    return v;
  }
  u32(): number {
    const v = this.view.getUint32(this.off);
    this.off += 4;
    return v;
  }
  u64(): number {
    const v = this.view.getBigUint64(this.off);
    this.off += 8;
    if (v > BigInt(Number.MAX_SAFE_INTEGER)) {
      throw new RangeError(`u64 value ${v} exceeds Number.MAX_SAFE_INTEGER`);
    }
    return Number(v);
  }
  string(): string {
    const len = this.u16();
    const s = TEXT_DECODER.decode(this.bytes.subarray(this.off, this.off + len));
    this.off += len;
    return s;
  }
  slice(len: number): Uint8Array {
    // Copy: the caller may hold the payload past buffer compaction.
    const out = this.bytes.slice(this.off, this.off + len);
    this.off += len;
    return out;
  }
}

/** Decode the payload bytes (after the tag byte) into a Frame. */
export function decodeFramePayload(tag: number, payload: Uint8Array): Frame {
  const c = new Cursor(new DataView(payload.buffer, payload.byteOffset, payload.byteLength), payload);

  switch (tag) {
    case FrameTag.NODE: {
      const nodeId = c.u32();
      const parentId = c.u32();
      const kindVal = c.u8();
      if (kindVal !== NodeKind.DIR && kindVal !== NodeKind.FILE) {
        throw new RangeError(`unknown node kind: ${kindVal}`);
      }
      const name = c.string();
      const sizeHint = c.u8() ? c.u64() : null;
      const mtimeMs = c.u8() ? c.u64() : null;
      return { tag: FrameTag.NODE, nodeId, parentId, kind: kindVal, name, sizeHint, mtimeMs };
    }
    case FrameTag.SUMMARY: {
      return {
        tag: FrameTag.SUMMARY,
        nodeId: c.u32(),
        scanDoneUnits: c.u64(),
        scanEstUnits: c.u64(),
        bodyDoneBytes: c.u64(),
        bodyEstBytes: c.u64(),
        sealed: c.u8() !== 0,
      };
    }
    case FrameTag.FILE_OPEN: {
      const nodeId = c.u32();
      const size = c.u64();
      const mtimeMs = c.u8() ? c.u64() : null;
      return { tag: FrameTag.FILE_OPEN, nodeId, size, mtimeMs };
    }
    case FrameTag.FILE_CHUNK: {
      const nodeId = c.u32();
      const offset = c.u64();
      const length = c.u32();
      const ckKind = c.u8();
      if (ckKind !== ChecksumKind.CRC32C) {
        throw new RangeError(`unknown checksum kind: ${ckKind}`);
      }
      const checksum = c.u32();
      const payloadBytes = c.slice(length);
      return {
        tag: FrameTag.FILE_CHUNK,
        nodeId,
        offset,
        length,
        checksumKind: ckKind,
        checksum,
        payload: payloadBytes,
      };
    }
    case FrameTag.FILE_CLOSE: {
      return { tag: FrameTag.FILE_CLOSE, nodeId: c.u32(), sizeSent: c.u64() };
    }
    case FrameTag.DIR_CLOSE: {
      return { tag: FrameTag.DIR_CLOSE, nodeId: c.u32() };
    }
    case FrameTag.SESSION_END: {
      return {
        tag: FrameTag.SESSION_END,
        scanDoneUnits: c.u64(),
        scanEstUnits: c.u64(),
        bodyDoneBytes: c.u64(),
        bodyEstBytes: c.u64(),
        sealed: c.u8() !== 0,
      };
    }
    case FrameTag.CLIENT_ABORT: {
      return { tag: FrameTag.CLIENT_ABORT, code: c.string(), reason: c.string() };
    }
    default:
      throw new RangeError(`unknown frame tag: 0x${tag.toString(16).padStart(2, "0")}`);
  }
}

/**
 * Incremental frame reader that buffers partial data from a streaming body.
 *
 * Feed chunks via `feed(data)` and collect decoded frames via `drain()`.
 *
 * `maxFrameLen` bounds the declared frame length: without it a corrupt or
 * malicious 4-byte prefix claiming a multi-gigabyte frame would grow the
 * buffer without limit while the reader waits for it to "complete".
 */
export class FrameReader {
  // 1 MiB default: comfortably above MAX_CHUNK_BYTES (256 KiB) + headers
  // and any NODE frame with a long UTF-8 name.
  static readonly DEFAULT_MAX_FRAME_LEN = 1024 * 1024;

  private chunks: Uint8Array[] = [];
  private buffered = 0;
  private readonly maxFrameLen: number;

  constructor(maxFrameLen: number = FrameReader.DEFAULT_MAX_FRAME_LEN) {
    this.maxFrameLen = maxFrameLen;
  }

  feed(data: Uint8Array): void {
    if (data.length > 0) {
      this.chunks.push(data);
      this.buffered += data.length;
    }
  }

  drain(): Frame[] {
    // Flatten once per drain; the partial tail is kept as a single chunk.
    const buf =
      this.chunks.length === 1 ? this.chunks[0] : concat(this.chunks, this.buffered);
    const view = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);

    const frames: Frame[] = [];
    let pos = 0;
    const total = buf.length;

    while (pos + 4 <= total) {
      const frameLen = view.getUint32(pos);
      if (frameLen > this.maxFrameLen) {
        throw new RangeError(`frame length ${frameLen} exceeds limit ${this.maxFrameLen}`);
      }
      if (pos + 4 + frameLen > total) break; // incomplete frame
      const tag = buf[pos + 4];
      // Copy payload so nothing pins the accumulation buffer.
      const payload = buf.slice(pos + 5, pos + 4 + frameLen);
      frames.push(decodeFramePayload(tag, payload));
      pos += 4 + frameLen;
    }

    if (pos > 0 || this.chunks.length > 1) {
      const tail = buf.subarray(pos);
      this.chunks = tail.length > 0 ? [tail] : [];
      this.buffered = tail.length;
    }

    return frames;
  }
}

function concat(chunks: Uint8Array[], total: number): Uint8Array {
  const out = new Uint8Array(total);
  let off = 0;
  for (const c of chunks) {
    out.set(c, off);
    off += c.length;
  }
  return out;
}

// ---------------------------------------------------------------------------
// CRC-32C (Castagnoli) — table-driven, same algorithm as @mfup/client.
//
// Unlike CPython (~6 MB/s pure-Python, hence the hard C dependency there),
// V8 runs this table loop at hundreds of MB/s — comfortably above any single
// upload stream. No native module needed.
// ---------------------------------------------------------------------------
const CRC32C_TABLE = new Uint32Array(256);
{
  for (let i = 0; i < 256; i++) {
    let crc = i;
    for (let j = 0; j < 8; j++) {
      crc = crc & 1 ? (crc >>> 1) ^ 0x82f63b78 : crc >>> 1;
    }
    CRC32C_TABLE[i] = crc >>> 0;
  }
}

export function crc32c(data: Uint8Array, initial = 0): number {
  let crc = (initial ^ 0xffffffff) >>> 0;
  for (let i = 0; i < data.length; i++) {
    crc = (crc >>> 8) ^ CRC32C_TABLE[(crc ^ data[i]) & 0xff];
  }
  return (crc ^ 0xffffffff) >>> 0;
}

export const CRC32C_IMPL = "js-table";
